docs: update conda instructions (#111)
lumicks/pylake: docs/install.rst

.. rubric:: Windows

#. Go to the `Anaconda`_ website and download the Python 3.8 (or newer) installer.

#. Run it and accept the default options during the installation.

#. Open `Anaconda Prompt` from the `Start` menu.

#. Create a new environment for Pylake named pylake by invoking the following command::

    conda create -n pylake

#. The environment can then be activated::

    conda activate pylake

> #. Verify that `(pylake)` is now prefixed to the path in your anaconda prompt (this means the environment is activated).

Reviewer: If we're targeting non-expert users, we might want to phrase this more something like:
> The line where you enter your next command should now start with the text `(pylake)` instead of `(base)`, to indicate you're in the _pylake_ environment you just created.

Author: Good point! Amended!

#. Pylake can be found on `conda-forge`_. Add `conda-forge`_ to the list of sources as follows::

    conda config --add channels conda-forge

#. We can install Pylake in this environment by invoking the following commands::

    conda install lumicks.pylake

#. To make sure Jupyter notebook can find the ipykernel, we invoke::

    python -m ipykernel install --user

#. It should be possible to open a jupyter notebook in this environment by calling::

>     jupyter notebook

Reviewer: These are useful steps! I didn't know. I always end up installing the Notebook in each and every environment, but this is much better 😄

Author: 👍

> Note that if you are used to starting Jupyter notebook from the anaconda navigator, you will have to set the environment to `pylake` for it to have access to the environment where you just installed Pylake.

Reviewer: ```suggestion
Note that if you are used to starting Jupyter notebook from the Anaconda Navigator, you will have to set the environment to `pylake` for it to have access to the environment where you just installed Pylake.
```

Author: Ah yes, I've capitalized it elsewhere now too.

You can do this in the drop down menu that normally defaults to the environment `base`.

> If you are used to starting Jupyter from the Anaconda prompt, then remember to activate the correct environment (pylake) prior to starting the notebook.

Reviewer: Might want to mention the actual `conda activate pylake` command again, just in case.

Author: Ah yes. I'll add it.

    conda update lumicks.pylake

> Note that this updates the package to the latest version that is compatible with your environment.

Reviewer: How about, after  listing the `conda update lumicks.pylake` command, we just say directly "if this gives errors, it's often easier to wipe the _pylake_ environment and re-install"? Then we list the following commands:
```
(Open a new Anaconda prompt)
conda env remove -n pylake
(Follow the installation instructions again)
```

I like that you mention this "nuclear option", as indeed it's usually the easiest way out of conflicts.

Author: One thing that is a bit of a pain is that it can quietly fail. It just won't update to the latest version. This is why the `conda list` call is in there. How about the following?

```rst
If you already have Pylake installed and you want to update to the latest version, just run::

    conda update lumicks.pylake

Note that this updates the package to the latest version that is compatible with your environment.
It will also attempt to update any dependencies that require an update in order to be compatible with the updated version of pylake.
You can check which version of pylake you have after this procedure by checking the pylake version
from the command prompt (windows) or terminal (macOS/linux)::

    conda list pylake

If for some reason conda fails to update pylake to the latest version, it is usually easier to just remove the pylake environment and reinstall from scratch.
To do this, open a new Anaconda prompt and type::

    conda env remove -n pylake

After which you can re-install pylake using the regular installation instructions above.
```

Reviewer: Yes, excellent!

Good comment about the `conda list` thing: I wasn't aware of that. (It's become a habit for me to always use clean environments, so I didn't observe the problem in my own use of _conda_.) Good to be aware of, and one more for the list of "annoying edge cases in _conda_". Here's to hoping that the Python packaging infrastructure will improve at some point 🙄

    conda install -c conda-forge ffmpeg

Conda environments
------------------
.. _PyCharm: https://www.jetbrains.com/pycharm/download/#section=windows
.. _PyCharm documentation: https://www.jetbrains.com/help/pycharm/conda-support-creating-conda-virtual-environment.html
.. _Jupyter: https://jupyter.org/
.. _uninstall instructions: https://docs.anaconda.com/anaconda/install/uninstall/

If you have installed Pylake according to the Installation Instructions for Anaconda, then you should now have a separate environment for your Pylake work.
You may be wondering why we needed to create a new environment for Pylake, and what an environment is.

When using Python, you will quickly find that several packages that you can install depend on each other.
For example, when installing Pylake, you could see that the installation of Pylake required numerous other packages to be installed as well.
Most Python packages are continuously updated by their authors. Sometimes, these authors decide that certain existing functionalities need to be changed.
This means that not all of them will be completely backwards compatible.

> Therefore, it can be challenging to find a suitable collection of packages that fulfills the requirements of the package you are trying to install, while ensuring that all packages are still compatible.

Reviewer: ```suggestion
Therefore, it can be challenging to find a set of packages and package versions that all work together.
```

Author: Ah yes, that sounds much better.

One pragmatic solution to this is to maintain separate Python environments for different projects.

> This means the solver doesn't have a large number of dependencies to resolve.

Reviewer: ```suggestion
This means that you create independent "copies" of Python and its installed packages, so that the different projects you're working on don't interfere with each other.
```
("solver" is not something you've defined; trying to rephase more in layman terms)

Author: Ah yes, much better. Amended! 👍

These installations are referred to as environments.

**Why do we install Pylake in a separate environment by default?**

Conda sources the packages it uses from a channel, these are locations where conda and the anaconda navigator search for packages.
The default one is called Anaconda, but Pylake is available on a channel named `conda-forge`_.
Conda forge and Anaconda both have different versions of different packages.
Some of these are not compatible with each other.
This is why it is wise to install Pylake into its own environment, and only source packages from the channel `conda-forge`_ in that environment.
This helps prevent difficulties when trying to come up with a plan to install a package you request.

**How do I set up my other tools to use the correct environment?**

For most programs, it is just a matter of pointing them to the correct environment.
If you prefer using the anaconda navigator, you can activate the environment by selecting it from the drop down menu `Applications on`.
By default, the selected environment is `base`.

> 

Reviewer: VS Code? 🥺 
For that one, the instructions would be: press Ctrl+Shift+P, and run the command "Python: Select Interpreter". It should automatically list all conda environments in the list that pops up.

Author: That's what I tried, but while that works right off the bat on Ubuntu, it doesn't seem to work on a clean VS Code install on Windows (I tried a clean windows image for this, to make sure it wasn't just my system). While I can see that it attempts to activate the environment in the powershell prompt that runs the python script, it seems that the environment is not actually correctly activated.

See this open issue here: https://github.com/numpy/numpy/issues/15183#issuecomment-603575874

We have two options. Either instruct them to always start VSCode from the Anaconda navigator (this works and activates the environment prior to starting VS Code). Instruct them to start VS Code from the anaconda prompt *after* activating the environment. Or instruct them to change `settings.json`, which will make it work no matter from where you start it. If it was my computer, the last would have my preference, but it is also the most complicated instruction. What do you think?

```rst
Go to the extensions tab (CTRL + SHIFT + X)

Enter Python in the search field.

Click on the Python plugin (by Microsoft) and install it.

Restart VSCode.

CTRL + Shift + P

Type Python and select the pylake environment from the list.

CTRL + SHIFT + P

Open settings

Add the following lines to settings.json:

    "terminal.integrated.shell.windows": "C:\\WINDOWS\\System32\\cmd.exe",
    "terminal.integrated.shellArgs.windows": ["/K", "x:\\xxx\\Anaconda3\\Scripts\\activate.bat x:\\xxx\\Anaconda3"],

In these lines, you have to replace `x:\\xxx` by the location of your Anaconda installation. You can find out where it is installed by starting an Anaconda prompt and typing `where activate.bat`.
```

Just for reference, this is the error you get when you don't:

```
PS D:\> conda activate test_pylake_latest
PS D:\>  & 'C:\Conda\envs\test_pylake_latest\python.exe' 'c:\Users\JoepVanlier\.vscode\extensions\ms-python.python-2021.1.502429796\pythonFiles\lib\python\debugpy\launcher' '63797' '--' 'd:\test.py'
Traceback (most recent call last):
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\numpy\core\__init__.py", line 22, in <module>  
    from . import multiarray
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\numpy\core\multiarray.py", line 12, in <module>
    from . import overrides
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\numpy\core\overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
ImportError: DLL load failed while importing _multiarray_umath: The specified module could not be found.   

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\test.py", line 1, in <module>
    import lumicks.pylake as lk
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\lumicks\pylake\__init__.py", line 4, in <module>
    from .file import *
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\lumicks\pylake\file.py", line 1, in <module>    
    import h5py
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\h5py\__init__.py", line 25, in <module>  
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\numpy\__init__.py", line 145, in <module>
    from . import core
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\numpy\core\__init__.py", line 48, in <module>
    raise ImportError(msg)
ImportError:

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.9 from "C:\Conda\envs\test_pylake_latest\python.exe"
  * The NumPy version is: "1.20.0"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: DLL load failed while importing _multiarray_umath: The specified module could not be found.

PS D:\>  d:; cd 'd:\\'; & 'C:\Conda\envs\test_pylake_latest\python.exe' 'c:\Users\JoepVanlier\.vscode\extensions\ms-python.python-2021.1.502429796\pythonFiles\lib\python\debugpy\launcher' '63882' '--' 'd:\test.py' 
Traceback (most recent call last):
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\numpy\core\__init__.py", line 22, in <module>
    from . import multiarray
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\numpy\core\multiarray.py", line 12, in <module>
    from . import overrides
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\numpy\core\overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
ImportError: DLL load failed while importing _multiarray_umath: The specified module could not be found.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\test.py", line 1, in <module>
    import lumicks.pylake as lk
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\lumicks\pylake\__init__.py", line 4, in <module>
    from .file import *
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\lumicks\pylake\file.py", line 1, in <module>
    import h5py
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\h5py\__init__.py", line 25, in <module>
    from . import _errors
  File "h5py\_errors.pyx", line 1, in init h5py._errors
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\numpy\__init__.py", line 145, in <module>
    from . import core
  File "C:\Conda\envs\test_pylake_latest\lib\site-packages\numpy\core\__init__.py", line 48, in <module>
    raise ImportError(msg)
ImportError:

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.9 from "C:\Conda\envs\test_pylake_latest\python.exe"
  * The NumPy version is: "1.20.0"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: DLL load failed while importing _multiarray_umath: The specified module could not be found.
```

Reviewer: Oh boy, such a mess. I understand your hesitation to include those instructions.

Perhaps we could give just the short instructions based on starting VS Code from the Navigator, and link to [this VS Code doc page](https://code.visualstudio.com/docs/python/environments#_conda-environments) (under the heading "Conda environments") for troubleshooting? We could even throw in that numpy discussion link as well for further troubleshooting. For novices, starting from the Navigator is probably just fine; and more experiences users can then easily figure the rest out themselves using the links provided.

What do you think?

Author: Yeah, referencing those pages is better. How about the following:

```rst
*VS Code*

#. Download `VS Code`_ and install it following the default installation instructions.

#. Start it when the installer finishes.

#. Go to the extensions tab (CTRL + SHIFT + X)

#. Enter Python in the search field.

#. Click on the Python plugin (by Microsoft) and install it.

#. Restart VS Code.

#. Open the Command Palette (CTRL + SHIFT + P) and type "Python: Select Interpreter". Here you should choose the pylake environment.

#. Close VS Code.

VS Code should now appear in your Anaconda Navigator list.
Make sure that you selected the pylake environment when starting VS Code from the Anaconda Navigator.
It should now be possible to use pylake in VS Code.

Under Windows, you will need to start VS Code from the Anaconda Navigator for it to use the correct environment.
For more information on how to get it to run without Anaconda Navigator on Windows see the `VS Code Environment Instructions`_ and this `numpy discussion`_.
```
